### DISCUSSION TITLE
authentication database?

### DISCUSSION THREAD

**Discussion by ssnyder** [pkosiec/mongo-seeding#159] [Q&A] [answered]
I'm interested in using this library for some of my automated tests.  My mongo db is setup with authentication enabled - using an authentication database, I don't see any documentation that would allow me to specify the equivalent of --authenticationDatabase .   Is it possible?

**Comment by pkosiec** [comment]
Hi @ssnyder, you can authenticate by using username and password parameters or by providing connectionURI. See the readme for particular solution (for example, if you want to use cli - see the readme: https://github.com/pkosiec/mongo-seeding/blob/master/cli/README.md). In JS library you can provide username and password for database object or DB connection uri. Does this answer your question?

PS sorry for mistakes, I'm writing this on my mobile because currently I don't have access to my computer.

EDIT:
I revisited this question and I just saw that I didn't tell that the `--authenticationDatabase` option is easily done with `DB_URI`. Here are the details, how to do it: https://dba.stackexchange.com/questions/185783/how-to-specify-authentication-database-and-target-database-separately-on-mongodb

**Comment by pkosiec** [comment]
I'm closing this for now, assuming that all questions have been answered. Feel free to reopen this and comment, if that's not true 🙂

**Comment by hongkongkiwi** [comment]
I would like to re-open this topic. It is possible to do this with a uri like this:
mongodb://test:testpwd@localhost/test?authSource=admin

This is necessary when a user is not using the default "admin" name for thier authentication database. For example, in my case I am using "users". I would still prefer to use ENV variables seperately rather than passing the whole URI so at present I cant use it this way.

To make it easy when using docker, can I suggest to add a new ENV called DB_MONGO_AUTH_DB. If it's not set, no need to do anything, but if is set, then you can simply add it as ?authSource=$DB_MONGO_AUTH_DB when constructing the uri in code.

I would suggest to change it like follows:

```typescript
  getDbConnectionUri({
    protocol,
    host,
    port,
    name,
    username,
    password,
    authDb,
  }: SeederDatabaseConfigObject) {
    let credentials = '';
    if (username) {
      credentials = `${username}${password ? `:${password}` : ''}@`;
    }

   let options = []
   if (authDb) {
       options.push(`authSource=${authDb}`)
   }

    if (protocol === 'mongodb+srv') {
      return `${protocol}://${credentials}${host}/${name}?${options.join('&')}`;
    }
    return `${protocol}://${credentials}${host}:${port}/${name}?${options.join('&')}`;
  }
```

Well, I guess in this format we could also pass an options array or string, but just to keep it simple authSource is the only option I need.

**Comment by pkosiec** [comment]
> It is possible to do this with a uri like this:
> mongodb://test:testpwd@localhost/test?authSource=admin

Yes, it is possible by providing your own DB_URI.

>  I would still prefer to use ENV variables seperately rather than passing the whole URI so at present I cant use it this way.
> To make it easy when using docker, can I suggest to add a new ENV called DB_MONGO_AUTH_DB. If it's not set, no need to do anything, but if is set, then you can simply add it as ?authSource=$DB_MONGO_AUTH_DB when constructing the uri in code.

Hmm, to be honest I don't want to support all possible options for DB URI with separate environmental variables. So maybe it would be better to have a generic env variable like `DB_OPTIONS`, when you can provide your own options in the way like: `DB_OPTIONS='replicaSet=test;connectTimeoutMS=300000'` to get `?replicaSet=test&connectTimeoutMS=300000` at the end of the URI. What do you think?

**Comment by hongkongkiwi** [comment]
That sounds like a very tidy and flexible solution 👍

**Comment by pkosiec** [comment]
Great - I created a new issue for adding this feature. It should be quite easy to add it, so it will come soon.

**Comment by pkosiec** [comment, marked as answer]
The functionality has been added as a part of #78 (PR: https://github.com/pkosiec/mongo-seeding/pull/80)
